lib/store/index.js

```diff
@@ -34,29 +40,74 @@ class Store {
   }
 
   @action
-  newKernel(kernel: Kernel) {
-    const mappedLanguage =
-      Config.getJson("languageMappings")[kernel.language] || kernel.language;
-    this.runningKernels.set(mappedLanguage, kernel);
+  newKernel(
+    kernel: Kernel,
+    filePath: string,
+    editor: atom$TextEditor,
+    grammar: atom$Grammar
+  ) {
+    if (isMultilanguageGrammar(editor.getGrammar())) {
+      this.kernelMapping.merge({ [filePath]: { [grammar.name]: kernel } });
+    } else {
+      this.kernelMapping.set(filePath, kernel);
+    }
+    const index = this.runningKernels.findIndex(k => k === kernel);
+    if (index === -1) {
+      this.runningKernels.push(kernel);
+    }
     // delete startingKernel since store.kernel now in place to prevent duplicate kernel
     this.startingKernels.delete(kernel.kernelSpec.display_name);
   }
```

Comment:
After this is merged, I can come back and :fire: this `startingKernels` hack. Having a file identifier in the store as we do now is a much cleaner way to prevent the same editor from starting duplicate kernels.